src-kmerge/KMerge/LoserTree.hs

```diff
@@ -35,69 +36,65 @@ placeholder = unsafeCoerce ()
-    setPrimArray ids 0 (len - 1) (-1)
     arr <- newSmallArray (len - 1) placeholder
-
-    loop ids arr (len - 1) xs0
+    case as of
+      [] -> do
+        sizeRef <- newPrimVar 0
+        holeRef <- newPrimVar 0
+        return $! (MLT sizeRef holeRef ids arr, a)
+      _ -> do
+        setPrimArray ids 0 (len - 1) (-1)
+        loop ids arr (len - 1) input
   where
-    !len = length xs0
+    !len = length input
 
-    loop :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
-    loop !_   !_   !_   []     = error "should not happen"
-    loop  ids  arr  idx (x:xs) = do
+    loop :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> NonEmpty a -> m (MutableLoserTree (PrimState m) a, a)
+    loop  ids  arr  idx (x :| xs) = do
         sift ids arr (parentOf idx) (parentOf idx) x idx xs
 
-    sift :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> Int -> a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
+    sift :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> Int -> a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, a)
     sift !ids !arr !idxX !j !x !idx0 xs = do
         !idxY <- readPrimArray ids j
         y     <- readSmallArray arr j
-        if idxY < 0
-        then do
+        case xs of
+          -- We should never call 'loop' with an empty list,
+          -- So only follow this logic branch if the list is non-empty!
+          e:es | idxY < 0 -> do
             writePrimArray  ids j idxX
             writeSmallArray arr j x
-            loop ids arr (idx0 + 1) xs
-        else if j <= 0
-        then do
-                if x <= y
-                then do
-                    sizeRef <- newPrimVar (len - 1)
-                    holeRef <- newPrimVar idxX
-                    return (MLT sizeRef holeRef ids arr, Just x)
-                else do
-                    writePrimArray  ids j idxX
-                    writeSmallArray arr j x
-                    sizeRef <- newPrimVar (len - 1)
-                    holeRef <- newPrimVar idxY
-                    return (MLT sizeRef holeRef ids arr, Just y)
-        else do
-                if x < y
-                then do
-                    sift ids arr idxX (parentOf j) x idx0 xs
-                else do
-                    writePrimArray  ids j idxX
-                    writeSmallArray arr j x
-                    sift ids arr idxY (parentOf j) y idx0 xs
-
-{-# SPECIALIZE newLoserTree :: forall a.   Ord a => [a] -> IO          (MutableLoserTree RealWorld a, Maybe a) #-}
-{-# SPECIALIZE newLoserTree :: forall a s. Ord a => [a] -> Strict.ST s (MutableLoserTree s         a, Maybe a) #-}
-{-# SPECIALIZE newLoserTree :: forall a s. Ord a => [a] -> Lazy.ST   s (MutableLoserTree s         a, Maybe a) #-}
+            loop ids arr (idx0 + 1) $ e :| es
```

Comment:
I'm not quite happy with this part. Based on the `error "should not happen"` we had before, it seems that if `idxY < 0` then we know that `xs` is non-empty. But if this invariant for some reason gets violated, an error might be nicer than continuing with the other branch, where then `holeRef` might point to something negative, which is probably much harder to debug.
So it might be nicer to leave the original structure of the `if idxY < 0` and then throw an error if `xs` is empty, e.g. using `NE.fromList` or `fromMaybe (error "...") . nonEmpty`. Ideally there would also be comments on why we expect this to hold, which (if I understand the code correctly) could be along the lines of this:

There is an invariant for `sift` that the amount of `-1` entries in `ids` is exactly `length xs`. Before every call to loop we overwrite a `-1`. Thus, if `idxY < 0`, there is a `-1` remaining, so `xs` can't be empty.